File: src/alignment.py
"""Local alignment — parent/child positioning + auto-bounded regions.

Components with a `parent` reference have their `pos` computed AFTER all
absolutely-positioned anchors are placed. Each child declares which side
of its parent it sits against (`align`) and how much breathing room
(`align_gap`).

Regions with a `contains` list have their `bounds` computed AFTER all
component positions are resolved — the bounding box automatically grows
to enclose every listed component INCLUDING its label area.

Resolution is depth-first with cycle detection; safe to call once per
diagram before rendering.

Example:

    Component("orch",       pos=(860, 200))
    Component("researcher", parent="orch", align="right",  align_gap=50)
    Component("planner",    parent="orch", align="bottom", align_gap=70)
    Component("fs",         parent="planner", align="right", align_gap=70)
    Component("todo",       parent="fs",      align="right", align_gap=20)

Moving `orch` by 50 px moves every descendant by 50 px automatically.
"""
from __future__ import annotations

from model import Component, Diagram, LABEL_HEIGHT


# Approximate per-char widths used to estimate label extent. The renderer
# does not have access to real font metrics; these constants are tuned for
# the diagram's font stack (14 px bold name, 11.5 px regular subtitle).
# Slight overestimate is intentional — narrow labels never overlap, wide
# labels get just-enough breathing room.
_NAME_CHAR_W = 7        # 14 px bold sans average ≈ 7 px/char
_SUB_CHAR_W  = 6        # 11.5 px regular sans ≈ 6 px/char


def _label_half_width(c: Component) -> int:
    """Approximate half-width of the longest text line under `c`'s icon.
    Used by the auto-layout resolver to prevent adjacent components' labels
    from overlapping. Returns 0 for shapes that don't render labels
    (annotation, badge)."""
    if c.shape in ("annotation", "badge"):
        return 0
    name_w = len(c.name) * _NAME_CHAR_W
    sub_w  = len(c.subtitle) * _SUB_CHAR_W
    return max(name_w, sub_w) // 2


def _effective_half(c: Component) -> tuple[int, int]:
    """`Component.half` widened to whichever is bigger: icon or label.
    Heights stay icon-only (`half[1]`); the label area below the icon is
    accounted for separately via `LABEL_HEIGHT`."""
    hw, hh = c.half
    return (max(hw, _label_half_width(c)), hh)


def resolve_alignments(diagram: Diagram) -> None:
    """Four-pass resolver:

    1. `_resolve_auto_layouts` — Figma-style: regions with `layout` direction
       position every child component along that axis with `gap` spacing.
    2. `_resolve_component_alignments` — pairwise parent/child anchoring
       (`align="right"` etc) for components not placed by auto-layout.
    3. `_resolve_region_bounds` — region bounding boxes computed from the
       enclosed components (after their positions are final).
    4. `_auto_size_canvas` — if `diagram.width`/`height` are 0, derive
       them from the resolved geometry (component + region + via extents)
       plus a margin. Explicit `canvas W x H` directive overrides.

    Mutates `diagram` in place."""
    _resolve_auto_layouts(diagram)
    _resolve_component_alignments(diagram)
    _resolve_region_bounds(diagram)
    _auto_size_canvas(diagram)


def _auto_size_canvas(diagram: Diagram, margin: int = 30) -> None:
    """Compute canvas dimensions from resolved geometry when not specified.

    Walks all components (icon + label extents), regions (rect bounds),
    and edge via points / explicit label positions. Computes both min and
    max extents — if the leftmost/topmost extent sits closer to the canvas
    edge than `margin`, every element is shifted to give it that margin.
    This lets orphan components (no region/layout) render correctly at
    their default `(0, 0)` position. Only overwrites `diagram.width` /
    `diagram.height` if they are 0 (sentinel)."""
    if diagram.width > 0 and diagram.height > 0:
        return

    min_x = min_y =  10**9
    max_x = max_y = -10**9

    for c in diagram.components:
        eff_hw = max(c.half[0], _label_half_width(c))
        lh     = LABEL_HEIGHT.get(c.shape, 0) if (c.name or c.subtitle) else 0
        left   = c.pos[0] - eff_hw
        right  = c.pos[0] + eff_hw
        top    = c.pos[1] - c.half[1]
        bottom = c.pos[1] + c.half[1] + lh
        if left   < min_x: min_x = left
        if top    < min_y: min_y = top
        if right  > max_x: max_x = right
        if bottom > max_y: max_y = bottom

    for r in diagram.regions:
        if r.bounds == (0, 0, 0, 0):
            continue                                 # invisible layout-only
        x, y, w, h = r.bounds
        if x     < min_x: min_x = x
        if y     < min_y: min_y = y
        if x + w > max_x: max_x = x + w
        if y + h > max_y: max_y = y + h

    for e in diagram.edges:
        for vx, vy in e.via:
            if vx < min_x: min_x = vx
            if vy < min_y: min_y = vy
            if vx > max_x: max_x = vx
            if vy > max_y: max_y = vy
        if e.label_pos is not None:
            lx, ly = e.label_pos
            if lx < min_x: min_x = lx
            if ly < min_y: min_y = ly
            if lx > max_x: max_x = lx
            if ly > max_y: max_y = ly

    if min_x > max_x:                                # nothing to size against
        return

    dx = margin - min_x if min_x < margin else 0
    dy = margin - min_y if min_y < margin else 0
    if dx or dy:
        for c in diagram.components:
            c.pos = (c.pos[0] + dx, c.pos[1] + dy)
        for r in diagram.regions:
            if r.bounds == (0, 0, 0, 0):
                continue
            x, y, w, h = r.bounds
            r.bounds = (x + dx, y + dy, w, h)
        for e in diagram.edges:
            e.via = [(vx + dx, vy + dy) for vx, vy in e.via]
            if e.label_pos is not None:
                e.label_pos = (e.label_pos[0] + dx, e.label_pos[1] + dy)
        max_x += dx
        max_y += dy

    if diagram.width == 0:
        diagram.width = max_x + margin
    if diagram.height == 0:
        diagram.height = max_y + margin

    # Enforce min 4:3 (landscape) aspect: pad width and re-center
    # horizontally so a tall vertical stack doesn't look like a column.
    # Wide canvases are left untouched.
    if diagram.width * 3 < diagram.height * 4:
        new_w = (diagram.height * 4) // 3
        shift = (new_w - diagram.width) // 2
        for c in diagram.components:
            c.pos = (c.pos[0] + shift, c.pos[1])
        for r in diagram.regions:
            if r.bounds == (0, 0, 0, 0):
                continue
            x, y, w, h = r.bounds
            r.bounds = (x + shift, y, w, h)
        for e in diagram.edges:
            e.via = [(vx + shift, vy) for vx, vy in e.via]
            if e.label_pos is not None:
                e.label_pos = (e.label_pos[0] + shift, e.label_pos[1])
        diagram.width = new_w


def _resolve_auto_layouts(diagram: Diagram) -> None:
    """For each region with `layout` set, position every contained component
    along the layout axis starting from the region's `pos` (top-left)."""
    for r in diagram.regions:
        if r.layout is None:
            continue
        if r.pos is None:
            raise ValueError(
                f"region {r.id!r} has layout={r.layout!r} but no `pos` anchor"
            )
        if not r.contains:
            continue

        children = [diagram.get(cid) for cid in r.contains]
        pad_x, pad_y = r.padding
        ox, oy = r.pos
        cursor_x = ox + pad_x
        cursor_y = oy + pad_y

        # Use EFFECTIVE half-width (max of icon and label) so adjacent
        # children's labels don't overlap. `gap` then measures the clear
        # distance between adjacent LABEL ends, not icon edges.
        effs = [_effective_half(c) for c in children]

        if r.layout == "horizontal":
            # Cross-axis: vertical alignment based on the tallest child icon
            max_h = max(eh for _, eh in effs)
            for c, (ew, eh) in zip(children, effs):
                _, ch = c.half
                if r.align == "start":
                    cy = cursor_y + ch
                elif r.align == "end":
                    cy = cursor_y + 2 * max_h - ch
                else:  # center (default)
                    cy = cursor_y + max_h
                c.pos = (cursor_x + ew, cy)
                cursor_x += ew * 2 + r.gap

        else:  # vertical
            # Cross-axis: horizontal alignment based on the widest child
            # (icon or label, whichever is wider).
            max_w = max(ew for ew, _ in effs)
            for c, (ew, eh) in zip(children, effs):
                cw, _ = c.half
                if r.align == "start":
                    cx = cursor_x + cw
                elif r.align == "end":
                    cx = cursor_x + 2 * max_w - cw
                else:  # center
                    cx = cursor_x + max_w
                c.pos = (cx, cursor_y + eh)
                cursor_y += eh * 2 + r.gap


def _resolve_component_alignments(diagram: Diagram) -> None:
    resolved: set[str] = set()

    def resolve(cid: str, path: tuple[str, ...] = ()) -> None:
        if cid in resolved:
            return
        if cid in path:
            chain = " → ".join((*path, cid))
            raise ValueError(f"alignment cycle: {chain}")

        comp = diagram.get(cid)
        if comp.parent is None:
            resolved.add(cid)
            return

        # Resolve the parent first (depth-first).
        resolve(comp.parent, path + (cid,))
        parent = diagram.get(comp.parent)

        if comp.align is None:
            raise ValueError(
                f"component {cid!r} has parent={comp.parent!r} but no align side"
            )

        comp.pos = _align_to(parent, comp, comp.align, comp.align_gap, comp.align_offset)
        resolved.add(cid)

    for c in diagram.components:
        resolve(c.id)


def _resolve_region_bounds(diagram: Diagram) -> None:
    """For each region with `contains`, compute its bounding box as the
    envelope of all listed components (icon + label area), padded.

    `padding_bottom` (when set on the region) overrides the default
    bottom padding (`padding[1]`); useful to compensate visually for the
    region label sitting ABOVE the rect."""
    for r in diagram.regions:
        if not r.contains:
            continue
        cells = [diagram.get(cid) for cid in r.contains]
        pad_x, pad_y = r.padding
        pad_b = r.padding_bottom if r.padding_bottom is not None else pad_y

        # Use effective half-width so the rect wraps labels too, not just
        # icons. Without this, long labels (e.g., "Academic Papers") would
        # extend past the region rect.
        eff_hws  = [max(c.half[0], _label_half_width(c)) for c in cells]
        xs_left  = [c.pos[0] - ew for c, ew in zip(cells, eff_hws)]
        xs_right = [c.pos[0] + ew for c, ew in zip(cells, eff_hws)]
        ys_top   = [c.pos[1] - c.half[1] for c in cells]
        # Bottom extent INCLUDES label area below the icon.
        ys_bot   = [c.pos[1] + c.half[1] + LABEL_HEIGHT.get(c.shape, 0)
                    for c in cells]

        x = min(xs_left)  - pad_x
        y = min(ys_top)   - pad_y
        w = max(xs_right) - x + pad_x
        h = max(ys_bot)   - y + pad_b
        r.bounds = (x, y, w, h)


def _align_to(parent: Component, child: Component, side: str,
              gap: int, offset: tuple[int, int]) -> tuple[int, int]:
    """Return the (cx, cy) for `child` aligned on `side` of `parent`."""
    px, py        = parent.pos
    p_hw, p_hh    = parent.half
    p_label       = LABEL_HEIGHT.get(parent.shape, 0)
    c_hw, c_hh    = child.half
    ox, oy        = offset

    match side:
        case "right":
            # parent's right edge + gap + child half-width
            cx = px + p_hw + gap + c_hw
            cy = py
        case "left":
            cx = px - p_hw - gap - c_hw
            cy = py
        case "bottom":
            # parent's bottom INCLUDING label area; child sits below that
            cx = px
            cy = py + p_hh + p_label + gap + c_hh
        case "top":
            cx = px
            cy = py - p_hh - gap - c_hh
        case _:
            raise ValueError(f"unknown align side: {side!r}")

    return (cx + ox, cy + oy)

File: packages/js/src/alignment.ts
/**
 * Local alignment — JS port of `packages/python/src/kymo/alignment.py`.
 *
 * The post-parse resolver that turns declared relationships into absolute
 * geometry. Six passes, run in order by {@link resolveAlignments}:
 *   1. auto-layouts        — Figma-style: regions with a `layout` direction
 *                            position every child along that axis.
 *   2. component alignments — pairwise parent/child anchoring (`@ parent side`).
 *   3. region bounds        — bounding boxes from the enclosed components.
 *   4. fan-in / fan-out     — stagger shared-anchor ports so arrows fan out.
 *   5. trunk lanes          — channel-route parallel Z-edges into lanes.
 *   6. auto-size canvas      — derive width/height + snap to an 8px grid.
 *
 * Independent implementation kept at parity with the Python algorithm.
 */
import {
  LABEL_HEIGHT, componentHalf, regionHalf, anchor, resolveAnchors, getComponent,
  type Component, type Diagram, type Edge, type Region, type Point, type Side,
} from "./model.js";

type Node = Component | Region;

const NAME_CHAR_W = 7;   // 14px bold sans ≈ 7 px/char
const SUB_CHAR_W = 6;    // 11.5px regular sans ≈ 6 px/char

/** Round half-to-even, matching Python's built-in `round()` — so grid snapping
 *  and lane offsets land on the same pixels as the Python implementation
 *  (`Math.round` differs only at exact .5 boundaries, where it rounds up). */
function pyRound(x: number): number {
  const f = Math.floor(x);
  const diff = x - f;
  if (diff < 0.5) return f;
  if (diff > 0.5) return f + 1;
  return f % 2 === 0 ? f : f + 1;   // exactly .5 → nearest even
}

function isRegion(n: Node): n is Region {
  return "bounds" in n;
}

function halfOf(n: Node): Point {
  return isRegion(n) ? regionHalf(n) : componentHalf(n);
}

/** Geometric centre of a node — components carry it as `pos`, regions derive
 *  it from `bounds` (Python sorts fan-in/out by the other endpoint's pos). */
function centerOf(n: Node): Point {
  return isRegion(n) ? anchor(n, "center") : n.pos;
}

function labelHalfWidth(c: Component): number {
  if (c.shape === "annotation" || c.shape === "badge") return 0;
  const nameW = c.name.length * NAME_CHAR_W;
  const subW = c.subtitle.length * SUB_CHAR_W;
  return Math.floor(Math.max(nameW, subW) / 2);
}

function effectiveHalf(c: Component): Point {
  const [hw, hh] = componentHalf(c);
  return [Math.max(hw, labelHalfWidth(c)), hh];
}

function isZeroBounds(b: [number, number, number, number]): boolean {
  return b[0] === 0 && b[1] === 0 && b[2] === 0 && b[3] === 0;
}

function buildNodeMap(diagram: Diagram): Map<string, Node> {
  const m = new Map<string, Node>();
  for (const r of diagram.regions) m.set(r.id, r);
  for (const c of diagram.components) m.set(c.id, c);   // components win on collision
  return m;
}

export function resolveAlignments(diagram: Diagram): void {
  resolveAutoLayouts(diagram);
  resolveComponentAlignments(diagram);
  resolveRegionBounds(diagram);
  staggerFaninEdges(diagram);
  staggerTrunkLanes(diagram);
  autoSizeCanvas(diagram);
}

// ── Pass 5: trunk-lane channel routing ─────────────────────────────────
function staggerTrunkLanes(diagram: Diagram, minStep = 8, maxStep = 16): void {
  const byId = buildNodeMap(diagram);
  type TrunkEntry = { e: Edge; sp: Point; dp: Point };
  const horiz = new Map<string, TrunkEntry[]>();
  const vert = new Map<string, TrunkEntry[]>();
  const push = (m: Map<string, TrunkEntry[]>, key: string, entry: TrunkEntry): void => {
    const arr = m.get(key);
    if (arr) arr.push(entry); else m.set(key, [entry]);
  };

  for (const e of diagram.edges) {
    if (e.via.length) continue;                        // explicit routing wins
    const src = byId.get(e.src);
    const dst = byId.get(e.dst);
    if (!src || !dst) continue;
    const [sa, da] = resolveAnchors(e, src, dst);
    const sp = anchor(src, sa);
    const dp = anchor(dst, da);
    if (sp[0] === dp[0] || sp[1] === dp[1]) continue;  // already axis-aligned
    if (sa === "left" || sa === "right") {
      push(horiz, `${pyRound(sp[0] / 8) * 8},${pyRound(dp[0] / 8) * 8}`, { e, sp, dp });
    } else {
      push(vert, `${pyRound(sp[1] / 8) * 8},${pyRound(dp[1] / 8) * 8}`, { e, sp, dp });
    }
  }

  const assign = (
    entries: TrunkEntry[],
    sortIdx: 0 | 1, channelWidth: number,
  ): void => {
    const n = entries.length;
    if (n <= 1) return;
    const overlaps = (t1: typeof entries[number], t2: typeof entries[number]): boolean => {
      const r1 = [t1.sp[sortIdx], t1.dp[sortIdx]].sort((a, b) => a - b);
      const r2 = [t2.sp[sortIdx], t2.dp[sortIdx]].sort((a, b) => a - b);
      return Math.max(r1[0], r2[0]) < Math.min(r1[1], r2[1]);
    };
    let any = false;
    for (let i = 0; i < n && !any; i++) {
      for (let j = i + 1; j < n; j++) {
        if (overlaps(entries[i], entries[j])) { any = true; break; }
      }
    }
    if (!any) return;
    const step = Math.max(minStep, Math.min(maxStep, Math.floor(channelWidth / (n + 1))));
    entries.sort((a, b) => (a.sp[sortIdx] - b.sp[sortIdx]) || (a.dp[sortIdx] - b.dp[sortIdx]));
    const mid = (n - 1) / 2;
    entries.forEach((t, i) => { t.e.trunkOffset = pyRound((i - mid) * step); });
  };

  for (const [key, entries] of horiz) {
    const [srcX, dstX] = key.split(",").map(Number);
    assign(entries, 1, Math.abs(dstX - srcX));
  }
  for (const [key, entries] of vert) {
    const [srcY, dstY] = key.split(",").map(Number);
    assign(entries, 0, Math.abs(dstY - srcY));
  }
}

// ── Pass 4: fan-in / fan-out port distribution ─────────────────────────
interface FanEntry {
  e: Edge;
  src: Node;
  dst: Node;
  sa: Side;
  da: Side;
}

function staggerFaninEdges(diagram: Diagram): void {
  const byId = buildNodeMap(diagram);
  const fanin = new Map<string, FanEntry[]>();
  const fanout = new Map<string, FanEntry[]>();
  const push = (m: Map<string, FanEntry[]>, key: string, entry: FanEntry): void => {
    const arr = m.get(key);
    if (arr) arr.push(entry); else m.set(key, [entry]);
  };

  for (const e of diagram.edges) {
    const src = byId.get(e.src);
    const dst = byId.get(e.dst);
    if (!src || !dst) continue;
    const [sa, da] = resolveAnchors(e, src, dst);
    const entry: FanEntry = { e, src, dst, sa, da };
    push(fanin, `${e.dst}|${da}`, entry);
    push(fanout, `${e.src}|${sa}`, entry);
  }

  const STEP = 16;

  const spread = (
    entriesIn: FanEntry[], anchorSide: Side, attr: "srcOffset" | "dstOffset",
    nodeOf: (t: FanEntry) => Node, otherOf: (t: FanEntry) => Node, minCount: number,
  ): void => {
    let entries = entriesIn;
    let n = entries.length;
    if (n < minCount) return;
    if (attr === "srcOffset") {
      entries = entries.filter((t) => !t.e.sharedPort);   // shared-port edges opt out
      if (entries.length < 1) return;
      n = entries.length;
    }
    const node = nodeOf(entries[0]);
    if (anchorSide === "left" || anchorSide === "right") {
      const crossSpan = halfOf(node)[1] * 2;
      entries.sort((a, b) => centerOf(otherOf(a))[1] - centerOf(otherOf(b))[1]);
      const spreadTotal = Math.min(crossSpan - 16, STEP * (n - 1));
      const mid = (n - 1) / 2;
      entries.forEach((t, i) => {
        const dy = pyRound(((i - mid) / Math.max(mid, 1)) * (spreadTotal / 2));
        const cur = t.e[attr];
        t.e[attr] = [cur[0], cur[1] + dy];
      });
    } else if (anchorSide === "top" || anchorSide === "bottom") {
      const crossSpan = halfOf(node)[0] * 2;
      entries.sort((a, b) => centerOf(otherOf(a))[0] - centerOf(otherOf(b))[0]);
      const spreadTotal = Math.min(crossSpan - 16, STEP * (n - 1));
      const mid = (n - 1) / 2;
      entries.forEach((t, i) => {
        const dx = pyRound(((i - mid) / Math.max(mid, 1)) * (spreadTotal / 2));
        const cur = t.e[attr];
        t.e[attr] = [cur[0] + dx, cur[1]];
      });
    }
  };

  for (const [key, entries] of fanin) {
    const da = key.split("|")[1] as Side;
    spread(entries, da, "dstOffset", (t) => t.dst, (t) => t.src, 2);
  }
  for (const [key, entries] of fanout) {
    const sa = key.split("|")[1] as Side;
    spread(entries, sa, "srcOffset", (t) => t.src, (t) => t.dst, 3);
  }
}

// ── Pass 6: auto-size canvas + snap to grid ────────────────────────────
function autoSizeCanvas(diagram: Diagram, margin = 30): void {
  if (diagram.width > 0 && diagram.height > 0) return;

  let minX = 1e9, minY = 1e9, maxX = -1e9, maxY = -1e9;

  for (const c of diagram.components) {
    const effHw = Math.max(componentHalf(c)[0], labelHalfWidth(c));
    const lh = (c.name || c.subtitle) ? (LABEL_HEIGHT[c.shape] ?? 0) : 0;
    minX = Math.min(minX, c.pos[0] - effHw);
    maxX = Math.max(maxX, c.pos[0] + effHw);
    minY = Math.min(minY, c.pos[1] - componentHalf(c)[1]);
    maxY = Math.max(maxY, c.pos[1] + componentHalf(c)[1] + lh);
  }
  for (const r of diagram.regions) {
    if (isZeroBounds(r.bounds)) continue;
    const [x, y, w, h] = r.bounds;
    minX = Math.min(minX, x); minY = Math.min(minY, y);
    maxX = Math.max(maxX, x + w); maxY = Math.max(maxY, y + h);
  }
  for (const e of diagram.edges) {
    for (const [vx, vy] of e.via) {
      minX = Math.min(minX, vx); minY = Math.min(minY, vy);
      maxX = Math.max(maxX, vx); maxY = Math.max(maxY, vy);
    }
    if (e.labelPos) {
      minX = Math.min(minX, e.labelPos[0]); minY = Math.min(minY, e.labelPos[1]);
      maxX = Math.max(maxX, e.labelPos[0]); maxY = Math.max(maxY, e.labelPos[1]);
    }
  }

  if (minX > maxX) return;   // nothing to size against

  const dx = minX < margin ? margin - minX : 0;
  const dy = minY < margin ? margin - minY : 0;
  if (dx || dy) {
    for (const c of diagram.components) c.pos = [c.pos[0] + dx, c.pos[1] + dy];
    for (const r of diagram.regions) {
      if (isZeroBounds(r.bounds)) continue;
      const [x, y, w, h] = r.bounds;
      r.bounds = [x + dx, y + dy, w, h];
    }
    for (const e of diagram.edges) {
      e.via = e.via.map(([vx, vy]) => [vx + dx, vy + dy] as Point);
      if (e.labelPos) e.labelPos = [e.labelPos[0] + dx, e.labelPos[1] + dy];
    }
    maxX += dx; maxY += dy;
  }

  if (diagram.width === 0) diagram.width = maxX + margin;
  if (diagram.height === 0) diagram.height = maxY + margin;

  // Enforce min 4:3 landscape aspect: pad width and re-centre horizontally.
  if (diagram.width * 3 < diagram.height * 4) {
    const newW = Math.floor((diagram.height * 4) / 3);
    const shift = Math.floor((newW - diagram.width) / 2);
    for (const c of diagram.components) c.pos = [c.pos[0] + shift, c.pos[1]];
    for (const r of diagram.regions) {
      if (isZeroBounds(r.bounds)) continue;
      const [x, y, w, h] = r.bounds;
      r.bounds = [x + shift, y, w, h];
    }
    for (const e of diagram.edges) {
      e.via = e.via.map(([vx, vy]) => [vx + shift, vy] as Point);
      if (e.labelPos) e.labelPos = [e.labelPos[0] + shift, e.labelPos[1]];
    }
    diagram.width = newW;
  }

  snapToGrid(diagram);
}

function snapToGrid(diagram: Diagram, grid = 8): void {
  const s = (v: number): number => pyRound(v / grid) * grid;
  const sUp = (v: number): number => Math.floor((v + grid - 1) / grid) * grid;

  for (const c of diagram.components) c.pos = [s(c.pos[0]), s(c.pos[1])];
  for (const e of diagram.edges) {
    e.via = e.via.map(([vx, vy]) => [s(vx), s(vy)] as Point);
    if (e.labelPos) e.labelPos = [s(e.labelPos[0]), s(e.labelPos[1])];
  }

  const byId = new Map(diagram.components.map((c) => [c.id, c]));
  for (const r of diagram.regions) {
    if (isZeroBounds(r.bounds) || !r.contains.length) continue;
    const cells = r.contains.map((cid) => byId.get(cid)).filter((c): c is Component => !!c);
    if (!cells.length) continue;
    const [padX, padY] = r.padding;
    const padB = r.paddingBottom ?? padY;
    const xsLeft = cells.map((c) => c.pos[0] - Math.max(componentHalf(c)[0], labelHalfWidth(c)));
    const xsRight = cells.map((c) => c.pos[0] + Math.max(componentHalf(c)[0], labelHalfWidth(c)));
    const ysTop = cells.map((c) => c.pos[1] - componentHalf(c)[1]);
    const ysBot = cells.map((c) =>
      c.pos[1] + componentHalf(c)[1] + ((c.name || c.subtitle) ? (LABEL_HEIGHT[c.shape] ?? 0) : 0));
    const x = Math.min(...xsLeft) - padX;
    const y = Math.min(...ysTop) - padY;
    r.bounds = [x, y, Math.max(...xsRight) - x + padX, Math.max(...ysBot) - y + padB];
  }

  diagram.width = sUp(diagram.width);
  diagram.height = sUp(diagram.height);
}

// ── Pass 1: Figma-style auto-layouts ───────────────────────────────────
function resolveAutoLayouts(diagram: Diagram): void {
  for (const r of diagram.regions) {
    const dir = r.layout as "horizontal" | "vertical" | null;
    if (dir == null) continue;
    if (r.pos == null) continue;
    if (!r.contains.length) continue;

    const children = r.contains.map((cid) => getComponent(diagram, cid));
    const [padX, padY] = r.padding;
    const [ox, oy] = r.pos;
    let cursorX = ox + padX;
    let cursorY = oy + padY;
    const effs = children.map(effectiveHalf);

    if (dir === "horizontal") {
      const maxH = Math.max(...effs.map(([, eh]) => eh));
      children.forEach((c, i) => {
        const [ew] = effs[i];
        const ch = componentHalf(c)[1];
        let cy: number;
        if (r.align === "start") cy = cursorY + ch;
        else if (r.align === "end") cy = cursorY + 2 * maxH - ch;
        else cy = cursorY + maxH;
        c.pos = [cursorX + ew, cy];
        cursorX += ew * 2 + r.gap;
      });
    } else {
      const maxW = Math.max(...effs.map(([ew]) => ew));
      children.forEach((c, i) => {
        const [, eh] = effs[i];
        const cw = componentHalf(c)[0];
        let cx: number;
        if (r.align === "start") cx = cursorX + cw;
        else if (r.align === "end") cx = cursorX + 2 * maxW - cw;
        else cx = cursorX + maxW;
        c.pos = [cx, cursorY + eh];
        cursorY += eh * 2 + r.gap;
      });
    }
  }
}

// ── Pass 2: parent/child alignment ─────────────────────────────────────
function resolveComponentAlignments(diagram: Diagram): void {
  const resolved = new Set<string>();

  const resolve = (cid: string, path: string[]): void => {
    if (resolved.has(cid)) return;
    if (path.includes(cid)) {
      throw new Error(`alignment cycle: ${[...path, cid].join(" → ")}`);
    }
    const comp = getComponent(diagram, cid);
    if (comp.parent === null) { resolved.add(cid); return; }

    resolve(comp.parent, [...path, cid]);
    const parent = getComponent(diagram, comp.parent);
    if (comp.align === null) {
      throw new Error(`component ${JSON.stringify(cid)} has parent=${JSON.stringify(comp.parent)} but no align side`);
    }
    comp.pos = alignTo(parent, comp, comp.align, comp.alignGap, comp.alignOffset);
    resolved.add(cid);
  };

  for (const c of diagram.components) resolve(c.id, []);
}

// ── Pass 3: region bounds from contents ────────────────────────────────
function resolveRegionBounds(diagram: Diagram): void {
  for (const r of diagram.regions) {
    if (!r.contains.length) continue;
    const cells = r.contains.map((cid) => getComponent(diagram, cid));
    const [padX, padY] = r.padding;
    const padB = r.paddingBottom ?? padY;
    const effHws = cells.map((c) => Math.max(componentHalf(c)[0], labelHalfWidth(c)));
    const xsLeft = cells.map((c, i) => c.pos[0] - effHws[i]);
    const xsRight = cells.map((c, i) => c.pos[0] + effHws[i]);
    const ysTop = cells.map((c) => c.pos[1] - componentHalf(c)[1]);
    const ysBot = cells.map((c) =>
      c.pos[1] + componentHalf(c)[1] + ((c.name || c.subtitle) ? (LABEL_HEIGHT[c.shape] ?? 0) : 0));
    const x = Math.min(...xsLeft) - padX;
    const y = Math.min(...ysTop) - padY;
    r.bounds = [x, y, Math.max(...xsRight) - x + padX, Math.max(...ysBot) - y + padB];
  }
}

function alignTo(
  parent: Component, child: Component, side: string, gap: number, offset: Point,
): Point {
  const [px, py] = parent.pos;
  const [pHw, pHh] = componentHalf(parent);
  const pLabel = LABEL_HEIGHT[parent.shape] ?? 0;
  const [cHw, cHh] = componentHalf(child);
  const [ox, oy] = offset;
  let cx: number, cy: number;
  switch (side) {
    case "right":  cx = px + pHw + gap + cHw; cy = py; break;
    case "left":   cx = px - pHw - gap - cHw; cy = py; break;
    case "bottom": cx = px; cy = py + pHh + pLabel + gap + cHh; break;
    case "top":    cx = px; cy = py - pHh - gap - cHh; break;
    default: throw new Error(`unknown align side: ${JSON.stringify(side)}`);
  }
  return [cx + ox, cy + oy];
}
